cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/artifact/RemoteArtifactRepositoryReader.java

```diff
@@ -92,6 +97,29 @@ public ArtifactDetail getArtifact(Id.Artifact artifactId) throws Exception {
     return artifactDetail;
   }
 
+  @Override
+  public InputStream getArtifactBytes(Id.Artifact artifactId) throws Exception {
+    HttpResponse httpResponse;
+    String namespaceId = artifactId.getNamespace().getId();
+    ArtifactScope scope = ArtifactScope.USER;
+    // Cant use 'system' as the namespace in the request because that generates an error, the namespace doesnt matter
+    // as long as it exists. Using default because it will always be there
+    if (ArtifactScope.SYSTEM.toString().equalsIgnoreCase(namespaceId)) {
+      namespaceId = "default";
+      scope = ArtifactScope.SYSTEM;
+    }
+    String url = String.format("namespaces/%s/artifacts/%s/versions/%s/download?scope=%s",
+                               namespaceId,
+                               artifactId.getName(),
+                               artifactId.getVersion(),
+                               scope);
+    HttpRequest.Builder requestBuilder = remoteClient.requestBuilder(HttpMethod.GET, url);
+    httpResponse = execute(requestBuilder.build());
+
+    return new ByteArrayInputStream(httpResponse.getResponseBody());
```

Comment:
This is bad. This load the whole artifact content in the memory. You should return an InputStream that streams the content from the HTTP response.

Reply:
done